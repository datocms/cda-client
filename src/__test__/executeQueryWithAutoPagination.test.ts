import { parse, print } from '@0no-co/graphql.web';
import { expect, test } from 'vitest';
import {
  convertToAutoPaginationQueryAndVariables,
  mergeAutoPaginationSplittedResults,
} from '../executeQueryWithAutoPagination';

test('convertToAutoPaginationQueryAndVariables: no variables, first > 100', () => {
  const query = parse(/* GraphQL */ `
    query BuildSitemapUrls {
      allBlogPosts {
        slug
      }
      entries: allSuccessStories(first: 2500) {
        ...SuccessStoryUrlFragment
      }
    }

    fragment SuccessStoryUrlFragment on SuccessStoryRecord {
      slug
    }
  `);

  const [newQuery, variables] = convertToAutoPaginationQueryAndVariables(query);
  expect([print(newQuery), variables]).toMatchSnapshot();
});

test('convertToAutoPaginationQueryAndVariables: no variables, first > 100, starting from item 14', () => {
  const query = parse(/* GraphQL */ `
    query BuildSitemapUrls {
      allBlogPosts {
        slug
      }
      entries: allSuccessStories(skip: 13, first: 526) {
        ...SuccessStoryUrlFragment
      }
    }

    fragment SuccessStoryUrlFragment on SuccessStoryRecord {
      slug
    }
  `);

  const [newQuery, variables] = convertToAutoPaginationQueryAndVariables(query);
  expect([print(newQuery), variables]).toMatchSnapshot();
});

test('convertToAutoPaginationQueryAndVariables: multiple selections to paginate', () => {
  const query = parse(/* GraphQL */ `
    query BuildSitemapUrls {
      allBlogPosts(first: 2500) {
        slug
      }
      entries: allSuccessStories(first: 2500) {
        ...SuccessStoryUrlFragment
      }
    }

    fragment SuccessStoryUrlFragment on SuccessStoryRecord {
      slug
    }
  `);

  expect(() =>
    convertToAutoPaginationQueryAndVariables(query),
  ).toThrowErrorMatchingSnapshot();
});

test('convertToAutoPaginationQueryAndVariables: no variables, first < 100', () => {
  const query = parse(/* GraphQL */ `
    query BuildSitemapUrls {
      entries: allSuccessStories(first: 50) {
        slug
      }
    }
  `);

  const [newQuery, variables] = convertToAutoPaginationQueryAndVariables(
    query,
    { other: 'foo' },
  );

  expect([print(newQuery), variables]).toMatchSnapshot();
});

test('convertToAutoPaginationQueryAndVariables: variable `first` arg, missing variables', () => {
  const query = parse(/* GraphQL */ `
    query BuildSitemapUrls($first: IntType!) {
      entries: allSuccessStories(first: $first) {
        ...SuccessStoryUrlFragment
      }
    }

    fragment SuccessStoryUrlFragment on SuccessStoryRecord {
      slug
    }
  `);

  expect(() =>
    convertToAutoPaginationQueryAndVariables(query),
  ).toThrowErrorMatchingSnapshot();
});

test('convertToAutoPaginationQueryAndVariables: variable `first` arg, with variables, `first` > 100', () => {
  const query = parse(/* GraphQL */ `
    query BuildSitemapUrls($first: IntType!) {
      entries: allSuccessStories(first: $first) {
        ...SuccessStoryUrlFragment
      }
    }

    fragment SuccessStoryUrlFragment on SuccessStoryRecord {
      slug
    }
  `);

  const [newQuery, variables] = convertToAutoPaginationQueryAndVariables(
    query,
    { first: 2500, other: 'foo' },
  );
  expect([print(newQuery), variables]).toMatchSnapshot();
});

test('convertToAutoPaginationQueryAndVariables: variable `first` arg, with variables, `first` < 100', () => {
  const query = parse(/* GraphQL */ `
    query BuildSitemapUrls($first: IntType!) {
      entries: allSuccessStories(first: $first) {
        slug
      }
    }
  `);

  const [newQuery, variables] = convertToAutoPaginationQueryAndVariables(
    query,
    { first: 50, other: 'foo' },
  );

  expect([print(newQuery), variables]).toMatchSnapshot();
});

test('mergeAutoPaginationSplittedResults: nothing to merge', () => {
  expect(
    mergeAutoPaginationSplittedResults({
      something: [
        {
          entries: [{ slug: 'foo' }],
        },
      ],
    }),
  ).toMatchSnapshot();
});

test('mergeAutoPaginationSplittedResults: splitted results to merge', () => {
  expect(
    mergeAutoPaginationSplittedResults({
      something: [
        {
          splitted_0_entries: [{ slug: 'foo' }],
          splitted_100_entries: [{ slug: 'bar' }],
          splitted_200_entries: [{ slug: 'qux' }],
        },
      ],
    }),
  ).toMatchSnapshot();
});
