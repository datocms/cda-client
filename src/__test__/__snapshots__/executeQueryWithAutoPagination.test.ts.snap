// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertToAutoPaginationQueryAndVariables: multiple selections to paginate 1`] = `[Error: Cannot auto-paginate multiple selections in a single query!]`;

exports[`convertToAutoPaginationQueryAndVariables: no variables, first < 100 1`] = `
[
  "query BuildSitemapUrls {
  entries: allSuccessStories(first: 50) {
    slug
  }
}",
  {
    "other": "foo",
  },
]
`;

exports[`convertToAutoPaginationQueryAndVariables: no variables, first > 100 1`] = `
[
  "query BuildSitemapUrls {
  allBlogPosts {
    slug
  }
  splitted_0_entries: allSuccessStories(first: 100, skip: 0) {
    ...SuccessStoryUrlFragment
  }
  splitted_100_entries: allSuccessStories(first: 100, skip: 100) {
    ...SuccessStoryUrlFragment
  }
  splitted_200_entries: allSuccessStories(first: 100, skip: 200) {
    ...SuccessStoryUrlFragment
  }
  splitted_300_entries: allSuccessStories(first: 100, skip: 300) {
    ...SuccessStoryUrlFragment
  }
  splitted_400_entries: allSuccessStories(first: 100, skip: 400) {
    ...SuccessStoryUrlFragment
  }
}

fragment SuccessStoryUrlFragment on SuccessStoryRecord {
  slug
}",
  {},
]
`;

exports[`convertToAutoPaginationQueryAndVariables: no variables, first > 100, starting from item 14 1`] = `
[
  "query BuildSitemapUrls {
  allBlogPosts {
    slug
  }
  splitted_13_entries: allSuccessStories(first: 100, skip: 13) {
    ...SuccessStoryUrlFragment
  }
  splitted_113_entries: allSuccessStories(first: 26, skip: 113) {
    ...SuccessStoryUrlFragment
  }
}

fragment SuccessStoryUrlFragment on SuccessStoryRecord {
  slug
}",
  {},
]
`;

exports[`convertToAutoPaginationQueryAndVariables: variable \`first\` arg, missing variables 1`] = `[Error: Expected variable 'first' to be a number]`;

exports[`convertToAutoPaginationQueryAndVariables: variable \`first\` arg, with variables, \`first\` < 100 1`] = `
[
  "query BuildSitemapUrls($first: IntType!) {
  entries: allSuccessStories(first: $first) {
    slug
  }
}",
  {
    "first": 50,
    "other": "foo",
  },
]
`;

exports[`convertToAutoPaginationQueryAndVariables: variable \`first\` arg, with variables, \`first\` > 100 1`] = `
[
  "query BuildSitemapUrls {
  splitted_0_entries: allSuccessStories(first: 100, skip: 0) {
    ...SuccessStoryUrlFragment
  }
  splitted_100_entries: allSuccessStories(first: 100, skip: 100) {
    ...SuccessStoryUrlFragment
  }
  splitted_200_entries: allSuccessStories(first: 100, skip: 200) {
    ...SuccessStoryUrlFragment
  }
  splitted_300_entries: allSuccessStories(first: 100, skip: 300) {
    ...SuccessStoryUrlFragment
  }
  splitted_400_entries: allSuccessStories(first: 100, skip: 400) {
    ...SuccessStoryUrlFragment
  }
}

fragment SuccessStoryUrlFragment on SuccessStoryRecord {
  slug
}",
  {
    "other": "foo",
  },
]
`;

exports[`mergeAutoPaginationSplittedResults: nothing to merge 1`] = `
{
  "something": [
    {
      "entries": [
        {
          "slug": "foo",
        },
      ],
    },
  ],
}
`;

exports[`mergeAutoPaginationSplittedResults: splitted results to merge 1`] = `
{
  "something": [
    {
      "entries": [
        {
          "slug": "foo",
        },
        {
          "slug": "bar",
        },
        {
          "slug": "qux",
        },
      ],
    },
  ],
}
`;
